pkg/solana/chainwriter/chain_writer.go

```diff
@@ -0,0 +1,324 @@
+package chainwriter
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"math/big"
+
+	"github.com/gagliardetto/solana-go"
+
+	commoncodec "github.com/smartcontractkit/chainlink-common/pkg/codec"
+	"github.com/smartcontractkit/chainlink-common/pkg/codec/encodings/binary"
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink-common/pkg/services"
+	"github.com/smartcontractkit/chainlink-common/pkg/types"
+
+	"github.com/smartcontractkit/chainlink-solana/pkg/solana/client"
+	"github.com/smartcontractkit/chainlink-solana/pkg/solana/codec"
+	"github.com/smartcontractkit/chainlink-solana/pkg/solana/fees"
+	"github.com/smartcontractkit/chainlink-solana/pkg/solana/txm"
+)
+
+const ServiceName = "SolanaChainWriter"
+
+type SolanaChainWriterService struct {
+	lggr   logger.Logger
+	reader client.Reader
+	txm    txm.TxManager
+	ge     fees.Estimator
+	config ChainWriterConfig
+
+	codecs map[string]types.Codec
+
+	services.StateMachine
+}
+
+var (
+	_ services.Service     = &SolanaChainWriterService{}
+	_ types.ContractWriter = &SolanaChainWriterService{}
+)
+
+// nolint // ignoring naming suggestion
+type ChainWriterConfig struct {
+	Programs map[string]ProgramConfig
+}
+
+type ProgramConfig struct {
+	Methods map[string]MethodConfig
+	IDL     string
+}
+
+type MethodConfig struct {
```

Comment:
wdyt about keeping configs separately similar to EVM Chain Components?

Reply:
What's the advantage of doing it that way?

Reply:
Its best practice, we have CR and multinode configs in pkg/solana. Configs should have unit tests and validations, so keeping them in a separate package is handy to have all of that in the same place. You can see [some examples here](https://github.com/smartcontractkit/chainlink-solana/tree/develop/pkg/solana/config).